### ISSUE TITLE
test-network: test_dhcp_client_ipv6_only and test_address_static are unstable

### ISSUE THREAD

**Issue by mrc0mmand** [systemd/systemd#29678]
In recent-ish CentOS CI I noticed a couple of reoccurring fails:

`test_dhcp_client_ipv6_only` (C8S only, as far as I know)
```
======================================================================
FAIL: test_dhcp_client_ipv6_only (__main__.NetworkdDHCPClientTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 5129, in test_dhcp_client_ipv6_only
    self.assertIn('DNS=2600::ee', output)
AssertionError: 'DNS=2600::ee' not found in '# This is private data. Do not parse.\nADMIN_STATE=configured\nOPER_STATE=routable\nCARRIER_STATE=carrier\nADDRESS_STATE=routable\nIPV4_ADDRESS_STATE=off\nIPV6_ADDRESS_STATE=routable\nONLINE_STATE=online\nREQUIRED_FOR_ONLINE=yes\nREQUIRED_OPER_STATE_FOR_ONLINE=degraded\nREQUIRED_FAMILY_FOR_ONLINE=any\nACTIVATION_POLICY=up\nNETWORK_FILE=/run/systemd/network/25-dhcp-client-ipv6-only.network\nNETWORK_FILE_DROPINS=""\nDNS=\nNTP=\nSIP=\nDOMAINS=\nROUTE_DOMAINS=\nLLMNR=yes\nMDNS=no\nDHCP6_CLIENT_IAID=0x7ec6b6c\nDHCP6_CLIENT_DUID=DUID-EN/Vendor:0000ab11617740de1342c3a20000\n'
```

Jobs:
  - https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-centos8/6903/artifact//systemd-centos-ci/artifacts_y8sai23g/testsuite-logs-upstream.vxK/systemd-networkd-tests.py_FAIL.log
  - https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-centos8/6889/artifact//systemd-centos-ci/artifacts_02zpmuyh/testsuite-logs-upstream.l8i/systemd-networkd-tests.py_FAIL.log

After comparing some test journals (with my limited network knowledge) it looks like the interface doesn't receive an RA with the 'MANAGED' flag set (for some reason), so the DHCPv6 setup is not done at all.

---

`test_address_static` (seen on both Arch Linux and C8S)

```
======================================================================
FAIL: test_address_static (__main__.NetworkdNetworkTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 2590, in test_address_static
    self.check_nftset('ifindex', 'dummy98')
  File "./test/test-network/systemd-networkd-tests.py", line 1031, in check_nftset
    self.assertRegex(output, r'.*elements = { [^}]*' + contents + r'[^}]* }.*')
AssertionError: Regex didn't match: '.*elements = { [^}]*dummy98[^}]* }.*' not found in 'table inet sd_test {\n\tset ifindex {\n\t\ttype iface_index\n\t}\n}'

======================================================================
FAIL: test_address_static (__main__.NetworkdNetworkTests.test_address_static)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/test/test-network/systemd-networkd-tests.py", line 2589, in test_address_static
    self.check_nftset('network4', r'10\.10\.1\.0/24')
  File "/build/test/test-network/systemd-networkd-tests.py", line 1031, in check_nftset
    self.assertRegex(output, r'.*elements = { [^}]*' + contents + r'[^}]* }.*')
AssertionError: Regex didn't match: '.*elements = { [^}]*10\\.10\\.1\\.0/24[^}]* }.*' not found in 'table inet sd_test {\n\tset network4 {\n\t\ttype ipv4_addr\n\t\tflags interval\n\t}\n}'
```

Jobs:
  - https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-vagrant-archlinux-sanitizers/6991/artifact//systemd-centos-ci/artifacts_all/artifacts_w5uxil5u/vagrant-logs.aV0/vagrant-arch-sanitizers-clang-testsuite.NdY/systemd-networkd-tests.py_FAIL.log
  - https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-cron-build/240/artifact//systemd-centos-ci/artifacts_all/artifacts_bzvfq6ns/testsuite-logs-upstream.Us4/systemd-networkd-tests.py_FAIL.log

Here the missing NFT sets weren't created, for some reason, but it apparently happens only sometimes, no idea why:

```
Oct 23 08:13:35 n27-10-168.pool.ci.centos.org systemd-networkd[250159]: Added NFT set: family inet, table sd_test, set addr4, IP address 10.10.1.1/24
Oct 23 08:13:35 n27-10-168.pool.ci.centos.org systemd-networkd[250159]: Added NFT set: family inet, table sd_test, set network4, IP address 10.10.1.1/24
Oct 23 08:13:35 n27-10-168.pool.ci.centos.org systemd-networkd[250159]: Failed to add NFT set: family inet, table sd_test, set ifindex, IP address 10.10.1.1/24, ignoring
```

```
Oct 19 16:50:11 archlinux systemd-networkd[376451]: Added NFT set: family inet, table sd_test, set addr4, IP address 10.10.1.1/24
Oct 19 16:50:11 archlinux systemd-networkd[376451]: Failed to add NFT set: family inet, table sd_test, set network4, IP address 10.10.1.1/24, ignoring
Oct 19 16:50:11 archlinux systemd-networkd[376451]: Failed to add NFT set: family inet, table sd_test, set ifindex, IP address 10.10.1.1/24, ignoring
```

/cc @yuwata

**Comment by yuwata** [comment]
cc @topimiettinen

**Comment by mrc0mmand** [comment]
As for the NFT stuff - there's also an occasional NFT-related TEST-75-RESOLVED fail, .e.g:

```
[ 3446.320510] testsuite-75.sh[49]: + nft flush ruleset
[ 3448.396208] testsuite-75.sh[49]: + nft flush ruleset
[ 3448.396208] testsuite-75.sh[49]: + nft add table inet sd_test
[ 3448.397752] testsuite-75.sh[49]: + nft add set inet sd_test c '{ type cgroupsv2; }'
[ 3448.399433] testsuite-75.sh[49]: + nft add set inet sd_test u '{ typeof meta skuid; }'
[ 3448.401447] testsuite-75.sh[49]: + nft add set inet sd_test g '{ typeof meta skgid; }'
[ 3448.403575] testsuite-75.sh[49]: + systemd-run -u test-nft.service -p DynamicUser=yes -p 'NFTSet=cgroup:inet:sd_test:c user:inet:sd_test:u group:inet:sd_test:g' sleep 10000
[ 3448.407808] systemd[1]: test-nft.service: Failed to load configuration: No such file or directory
[ 3448.408293] systemd[1]: test-nft.service: Trying to enqueue job test-nft.service/start/fail
[ 3448.408321] systemd[1]: Added job test-nft.service/start to transaction.
[ 3448.408352] systemd[1]: Pulling in sysinit.target/start from test-nft.service/start
[ 3448.413611] systemd[1]: Pulling in system.slice/start from test-nft.service/start
[ 3448.413636] systemd[1]: Pulling in tmp.mount/start from test-nft.service/start
[ 3448.413662] systemd[1]: Pulling in shutdown.target/stop from test-nft.service/start
[ 3448.415809] systemd[1]: test-nft.service: Installed new job test-nft.service/start as 386
[ 3448.415835] systemd[1]: test-nft.service: Enqueued job test-nft.service/start as 386
[ 3448.416817] systemd[1]: test-nft.service: Will spawn child (service_enter_start): /usr/bin/sleep
[ 3448.416842] systemd[1]: test-nft.service: Failed to set 'memory.zswap.max' attribute on '/system.slice/test-nft.service' to 'max': No such file or directory
[ 3448.416867] systemd[1]: Added NFT set: family inet, table sd_test, set c, cgroup 76139
[ 3448.428183] systemd[1]: Failed to add NFT set: family inet, table sd_test, set u, cgroup 76139, ignoring: Invalid argument
[ 3448.444279] systemd[1]: Failed to add NFT set: family inet, table sd_test, set g, cgroup 76139, ignoring: Invalid argument
[ 3448.444466] systemd[1]: test-nft.service: Passing 0 fds to service
[ 3448.444499] systemd[1]: test-nft.service: About to execute: /usr/bin/sleep 10000
[ 3448.444751] systemd[1]: test-nft.service: Forked /usr/bin/sleep as 861
[ 3448.445372] systemd[1]: test-nft.service: Changed dead -> running
[ 3448.445397] systemd[1]: test-nft.service: Job 386 test-nft.service/start finished, result=done
[ 3448.446540] testsuite-75.sh[860]: Running as unit: test-nft.service; invocation ID: 7f84b646a44946db800ceff6efafc295
[ 3448.447122] testsuite-75.sh[49]: + run nft list set inet sd_test c
[ 3448.448892] testsuite-75.sh[862]: + nft list set inet sd_test c
[ 3448.451858] testsuite-75.sh[863]:                 elements = { "system.slice/test-nft.service" }
[ 3448.452977] testsuite-75.sh[49]: + grep -qF test-nft.service /tmp/tmp.KmXUkAqDNR
[ 3448.454379] testsuite-75.sh[866]: ++ getent passwd test-nft
[ 3448.455704] systemd[1]: varlink-12: New incoming message: {"method":"io.systemd.UserDatabase.GetUserRecord","parameters":{"userName":"test-nft","service":"io.systemd.DynamicUser"}}
[ 3448.457132] systemd[1]: testsuite-75.service: Failed with result 'exit-code'.
```

Journal: https://mrc0mmand.fedorapeople.org/TEST-75-RESOLVED-NFT.journal

**Comment by topimiettinen** [comment]
I ran `test-nft-set add inet sd_test c uint64 1234; test-nft-set del inet sd_test c uint64 1234` 10000 times without errors. This exercises the kernel interface with `nft_set_element_modify_any()`, it's the same what networkd and PID1 use. I don't think it's possible for networkd or PID1 to pass invalid data to the function, which would warrant kernel responding with EINVAL. Perhaps there's a problem with the kernel, especially if this is a recent issue.

Another suggestion could be examining `sd_nfnl_call_batch()`. At some point of development I saw problems with asynchronous mode of a fresh netlink socket, where the kernel would respond with EINPROGRESS. I don't think that's related but could there be other similar problems?

**Comment by mrc0mmand** [comment]
FWIW, I can reproduce the `Failed to add NFT set` messages from TEST-75 with a simple reproducer:

```bash
#!/bin/bash

set -eux
set -o pipefail

nft flush ruleset
nft add table inet sd_test
nft add set inet sd_test c '{ type cgroupsv2; }'
nft add set inet sd_test u '{ typeof meta skuid; }'
nft add set inet sd_test g '{ typeof meta skgid; }'

while :; do
    TS="$(date +"%s")"
    systemd-run -u test-nft.service -p DynamicUser=yes -p 'NFTSet=cgroup:inet:sd_test:c user:inet:sd_test:u group:inet:sd_test:g' true
    journalctl --sync
    journalctl --since "@$TS" --grep "Failed to add NFT" && break
    sleep 1
done
```
```
# ./repro.sh 
+ set -o pipefail
+ nft flush ruleset
+ nft add table inet sd_test
+ nft add set inet sd_test c '{ type cgroupsv2; }'
+ nft add set inet sd_test u '{ typeof meta skuid; }'
+ nft add set inet sd_test g '{ typeof meta skgid; }'
+ :
++ date +%s
+ TS=1698178369
+ systemd-run -u test-nft.service -p DynamicUser=yes -p 'NFTSet=cgroup:inet:sd_test:c user:inet:sd_test:u group:inet:sd_test:g' true
Running as unit: test-nft.service; invocation ID: 42f6a59026554479805a072446030ebd
+ journalctl --sync
+ journalctl --since @1698178369 --grep 'Failed to add NFT'
Oct 24 20:12:49 arch systemd[1]: Failed to add NFT set: family inet, table sd_test, set u, cgroup 4272, ignoring: Invalid argument
Oct 24 20:12:49 arch systemd[1]: Failed to add NFT set: family inet, table sd_test, set g, cgroup 4272, ignoring: Invalid argument
+ break
```

I'll try to come up with something for the networkd part tomorrow.

**Comment by mrc0mmand** [comment]
Nevermind, here's there networkd reproducer:

```bash
#!/bin/bash

set -eux
set -o pipefail

nft flush ruleset
nft add table inet sd_test
nft add set inet sd_test c '{ type cgroupsv2; }'
nft add set inet sd_test u '{ typeof meta skuid; }'
nft add set inet sd_test g '{ typeof meta skgid; }'

while :; do
    TS="$(date +"%s")"
    systemd-run -u test-nft.service -p DynamicUser=yes -p 'NFTSet=cgroup:inet:sd_test:c user:inet:sd_test:u group:inet:sd_test:g' true
    journalctl --sync
    journalctl --since "@$TS" --grep "Failed to add NFT" && break
    sleep 1
done
```

```
# ./repro2.sh 
+ set -o pipefail
+ nft flush ruleset
+ nft add table inet sd_test
+ nft add set inet sd_test addr6 '{' type 'ipv6_addr;' '}'
+ nft add set inet sd_test network6 '{' type 'ipv6_addr;' flags 'interval;' '}'
+ nft add set inet sd_test ifindex '{' type 'iface_index;' '}'
+ ip link add dev dummy99 type dummy
RTNETLINK answers: File exists
+ :
+ mkdir -p /run/systemd/network
+ cat
+ networkctl reload
+ :
++ date +%s
+ TS=1698180390
+ systemctl restart systemd-networkd
+ sleep 5
+ journalctl --sync
+ journalctl --since @1698180390 --grep 'Failed to add NFT'
Oct 24 20:46:31 arch systemd-networkd[1231]: Failed to add NFT set: family inet, table sd_test, set addr4, IP address 10.10.1.1/24, ignoring
Oct 24 20:46:31 arch systemd-networkd[1231]: Failed to add NFT set: family inet, table sd_test, set network4, IP address 10.10.1.1/24, ignoring
+ break
```

**Comment by mrc0mmand** [comment]
@yuwata looks like the first issue is still happening :(

```
======================================================================
FAIL: test_dhcp_client_ipv6_only (__main__.NetworkdDHCPClientTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/test-network/systemd-networkd-tests.py", line 5138, in test_dhcp_client_ipv6_only
    self.assertIn('DNS=2600::ee', output)
AssertionError: 'DNS=2600::ee' not found in '# This is private data. Do not parse.\nADMIN_STATE=configured\nOPER_STATE=routable\nCARRIER_STATE=carrier\nADDRESS_STATE=routable\nIPV4_ADDRESS_STATE=off\nIPV6_ADDRESS_STATE=routable\nONLINE_STATE=online\nREQUIRED_FOR_ONLINE=yes\nREQUIRED_OPER_STATE_FOR_ONLINE=degraded\nREQUIRED_FAMILY_FOR_ONLINE=any\nACTIVATION_POLICY=up\nNETWORK_FILE=/run/systemd/network/25-dhcp-client-ipv6-only.network\nNETWORK_FILE_DROPINS=""\nDNS=\nNTP=\nSIP=\nDOMAINS=\nROUTE_DOMAINS=\nLLMNR=yes\nMDNS=no\nDHCP6_CLIENT_IAID=0x7ec6b6c\nDHCP6_CLIENT_DUID=DUID-EN/Vendor:0000ab11617740de1342c3a20000\n'
```

https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-centos8/6986/artifact//systemd-centos-ci/index.html

**Comment by topimiettinen** [comment]
I think I found the problem for the NFT set errors. Size of NFT types for UIDs and GIDs is 32 bits but for cgroup IDs its 64 bits. The missing check (fixed with #29715) caused PID1 to try to push cgroup IDs to UID and GID sets with 64 bit element size, which wasn't accepted by kernel and thus there's EINVAL response.

**Comment by mrc0mmand** [comment]
I haven't seen any systemd-networkd-tests related fail since the three linked PRs were merged, so I guess we can close this.
